src/BalanceTx.purs

```diff
@@ -0,0 +1,611 @@
+module BalanceTx
+  ( balanceTxM
+  , UnbalancedTransaction(..)
+  )
+  where
+
+import Prelude
+import Data.Array ((\\), findIndex, modifyAt)
+import Data.Array as Array
+import Data.BigInt (BigInt, fromInt, quot)
+import Data.Either (Either(..), hush, note)
+import Data.Foldable as Foldable
+import Data.List ((:), List(..), partition)
+import Data.Map as Map
+import Data.Maybe (fromMaybe, Maybe(..))
+import Data.Newtype (unwrap, wrap)
+import Data.String.CodeUnits (length)
+import Data.Tuple (fst)
+import Data.Tuple.Nested ((/\), type (/\))
+import Undefined (undefined)
+
+import Ogmios (QueryM)
+import ProtocolParametersAlonzo (coinSize, lovelacePerUTxOWord, pidSize, protocolParamUTxOCostPerWord, utxoEntrySizeWithoutVal)
+import Types.Ada (adaSymbol, fromValue, getLovelace, lovelaceValueOf)
+import Types.Transaction (Address, Credential(..), RequiredSigner, Transaction(..), TransactionInput, TransactionOutput(..), TxBody(..), Utxo, UtxoM)
+import Types.Value (allTokenNames, emptyValue, flattenValue, geq, getValue, isAdaOnly, isPos, isZero, minus, numCurrencySymbols, numTokenNames, TokenName, Value(..))
+
+-- This module replicates functionality from
+-- https://github.com/mlabs-haskell/mlabs-pab/blob/master/src/MLabsPAB/PreBalance.hs
+
+-- Output utxos with the amount of lovelaces required.
+type MinUtxos = Array (TransactionOutput /\ BigInt)
+
+-- TO DO: convert utxosAt from Ogmios to Transaction space.
+utxosAt :: Address -> QueryM UtxoM
+utxosAt = undefined
+
+newtype PubKeyHash = PubKeyHash String
+newtype PrivateKey = PrivateKey String
+
+-- getPkhFromAddress :: Address -> PubKeyHash
+-- getPkhFromAddress = undefined
+
+getPrivKeys :: QueryM (Map.Map Address PrivateKey)
+getPrivKeys = undefined
+
+-- Taken from https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-constraints/html/Ledger-Constraints-OffChain.html#t:UnbalancedTx
+-- to demonstrate what is required. This is temporary to make the compiler happy.
+-- I haven't copied it exactly, instead just a minimal version given our
+-- incomplete Transaction.purs types. Also requiredSignatures is already in
+-- TxBody. I've ignored validity interval too.
+newtype UnbalancedTransaction = UnbalancedTransaction
+  { unbalancedTx :: Transaction
+  , utxoIndex :: UtxoM
+    -- utxoIndex will include utxos for scripts as well presumably.
+  }
+
+-- https://github.com/mlabs-haskell/mlabs-pab/blob/master/src/MLabsPAB/PreBalance.hs#54
+-- | Balances an unbalanced transaction
+balanceTxM
+  :: Address
+  -> UnbalancedTransaction
+  -> QueryM (Either String Transaction)
```

Comment:
I think it might be a good idea to make a more specific error type (applies to the whole module)

Reply:
Great shout.